snakemake/script.py

```diff
@@ -381,42 +410,112 @@ def generate_preamble(
         shadow_dir,
         preamble_addendum="",
     ):
-        wrapper_path = path[7:] if path.startswith("file://") else path
-        snakemake = Snakemake(
-            input_,
-            output,
-            params,
-            wildcards,
-            threads,
-            resources,
-            log,
-            config,
-            rulename,
-            bench_iteration,
-            os.path.dirname(wrapper_path),
-        )
-        snakemake = pickle.dumps(snakemake)
-        # Obtain search path for current snakemake module.
-        # The module is needed for unpickling in the script.
-        # We append it at the end (as a fallback).
+
+        # Obtain search path for any mounted containers
+        # This will also ensure recipes that require snakemake have it on path
         searchpath = SNAKEMAKE_SEARCHPATH
         if container_img is not None:
             searchpath = singularity.SNAKEMAKE_MOUNTPOINT
         searchpath = repr(searchpath)
+
         # For local scripts, add their location to the path in case they use path-based imports
         if path.startswith("file://"):
             searchpath += ", " + repr(os.path.dirname(path[7:]))
 
+        # The wrapper path should be added too
+        wrapper_path = path[7:] if path.startswith("file://") else path
+
         return textwrap.dedent(
             """
         ######## snakemake preamble start (automatically inserted, do not edit) ########
-        import sys; sys.path.extend([{searchpath}]); import pickle; snakemake = pickle.loads({snakemake}); from snakemake.logging import logger; logger.printshellcmds = {printshellcmds}; {preamble_addendum}
+        import sys, os; sys.path.extend([{searchpath}]); sys.path.insert(0, {sourcedir});from collections import namedtuple
```

Comment:
@johanneskoester here is what we talked about, removing snakemake and the picking. This isn't going to work for a lot of wrappers that use snakemake.shell, but it should work for future wrappers that don't. If the snakemake path that is added doesn't exist, I don't think it should hurt. But if it does exist, snakemake should be found akin to before.

Reply:
Ok, so this PR needs to be combined with a PR on snakemake-wrappers that gets rid of snakemake.shell. `os.system(f"command with {flags}")` is equally readable.